### PR TITLE
[Metricbeat][Aerospike Module] Add support for TLS

### PR DESCRIPTION
**Pull request by herrBez**
## Proposed commit message

[Aerospike] Add TLS support

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Aerospike has a documentation https://aerospike.com/docs/server/guide/security/tls documentation that explains how to configure aerospike with TLS.

Nevertheless, in my tests I used NGINX with the following configuration to perform SSL termination:
```
user  nginx;
worker_processes  auto;

events {
}

stream {
    server {
       listen 4333 ssl;
       proxy_pass stream_aerospike;
       ssl_certificate     /etc/nginx/certs/remote_fake.crt;
       ssl_certificate_key /etc/nginx/certs/remote_fake.key;
       ssl_client_certificate /etc/nginx/certs/ca.crt;

     }

     upstream stream_aerospike {
         # hash $remote_addr consistent;
         server aerospike:3000;
     }
}
```
And the following docker-compose.yml:
```yml
version: '2.3'

services:
  aerospike:
    image: docker.elastic.co/integrations-ci/beats-aerospike:${AEROSPIKE_VERSION:-3.9.0}-1
    build:
      context: ./_meta
      args:
        AEROSPIKE_VERSION: ${AEROSPIKE_VERSION:-3.9.0}
    ports:
      - 3000
  reverseproxy:
    image: nginx:latest
    volumes:
      - ./nginx/certs:/etc/nginx/certs
      - ./nginx/nginx.conf:/etc/nginx/nginx.conf  
    ports:
      - 4333:4333
``` 

## Related issues

- https://discuss.elastic.co/t/configuring-aerospike-metricbeat-module-with-tls/294729 

## Use cases

Monitor TLS and Username/Password protected Aerospike Databases

### CODE REVIEW

File: metricbeat/module/aerospike/aerospike_test.go
Comment (herrBez) [inline metricbeat/module/aerospike/aerospike_test.go:26]: Not sure what does it mea, any guidance is much appreciated

File: metricbeat/module/aerospike/aerospike_test.go
Comment (herrBez) [inline metricbeat/module/aerospike/aerospike_test.go:121]: I guess the problem is that tls.Config is initialized with default values (and it probably includes TLSv1.0). Is there a constructor that I can use to get a "default" SSL Configuration for Metricbeat.
 
 I can only think about, changing this piece of code as follows
 ```go
 	tlsconfig, _ := tlscommon.LoadTLSConfig(&tlscommon.Config{Enabled: getBoolPointer(true)})
	TLSPolicy.TlsConfig = tlsconfig.ToConfig()
 ```

File: metricbeat/modules.d/aerospike.yml.disabled
Comment (herrBez) [inline metricbeat/modules.d/aerospike.yml.disabled:1]: Can someone provide some guidance on how to actually upgrade/modify the file? `mage update` is failing. 

If changing this file it's not desired or considered out-of-scope, I can revert the change.

File: metricbeat/module/aerospike/aerospike_test.go
Comment (fearful-symmetry) [reply metricbeat/module/aerospike/aerospike_test.go:25]: The imports starting with `github.com/elastic/` need to appear before other 3rd party imports, run `goimports -w -local github.com/elastic metricbeat/module/aerospike`

File: metricbeat/modules.d/aerospike.yml.disabled
Comment (fearful-symmetry) [reply metricbeat/modules.d/aerospike.yml.disabled:1]: So, it's a little counter-intuitive, you need to make the changes in `_meta/config.yml` and then run `mage update`.

Comment (herrBez) [reply metricbeat/modules.d/aerospike.yml.disabled:1]: Thank you very much for the help!

File: metricbeat/module/aerospike/aerospike_test.go
Comment (herrBez) [reply metricbeat/module/aerospike/aerospike_test.go]: Thank you. Let's see if it's fixed now :)

**Comment by fearful-symmetry** [comment]
@herrBez so, I don't have much experience with Aerospike---is the idea there that basic auth would be used with the access controls in enterprise, or with a reverse proxy like nginx? If it's the latter (or both) it might be good to comment that and provide an example.

File: metricbeat/module/aerospike/aerospike_test.go
Comment (leehinman) [inline metricbeat/module/aerospike/aerospike_test.go:196]: would you mind switching to `Equalf` and providing a description of what didn't match?  

Right now if this fails in CI we will get an error like `"A" does not equal "B"`.  It is a lot easier to debug if we get something like `Aerospike policy username is wrong. got: "A" expected: "B"`

**Review by leehinman — CHANGES_REQUESTED** [review]

File: metricbeat/module/aerospike/aerospike_test.go
Comment (herrBez) [reply metricbeat/module/aerospike/aerospike_test.go:196]: Hi,
Thank you for the feedback. I don't mind at all ^^. So something like:

```go
assert.Equalf(t, test.expectedClientPolicy.User, result.User, "Aerospike policy [Test '%s'] username is wrong: expected '%s' got '%s'", test.Name, test.expectedClientPolicy.User, result.User)
```

Or something like:
```go
assert.Equalf(t, test.expectedClientPolicy.User, result.User, "Aerospike policy [Test '%s'] username is wrong", test.Name, test.expectedClientPolicy.User, result.User)
```
??

Should I also modify the existing tests to use Equalf?

Comment (leehinman) [reply metricbeat/module/aerospike/aerospike_test.go:196]: I don't think you need the test name, when it errors out go test will give us that.  I prefer the ones that have expected and got (or something like that).

One thing I think that helps is to fail the test and just double check that the error message gives you enough context that you have some idea of where the problem is.  For example telling me that the Username is what didn't match, the value we got, and the expected value goes a long way to helping understand where the problem might be.

I think you can just focus on the tests you are adding.  That helps limit the scope of the PR.

**Comment by herrBez** [comment]
After some digging/testing I think I will reduce the scope of this PR only to TLS/SSL support rather than adding also basic auth. Adding support for Basic-Auth (in a forward-compatible manner) would require an upgrade of the dependencies as well (I think at least [v.1.35.1 of the Go-Client](https://download.aerospike.com/download/client/go/notes.html#1.35.1)), which would make the context of this PR bigger than I imagined. By the way according to vendor's website only few versions of Aerospike DB are supported https://aerospike.com/docs/reference/platform-support#aerospike-database-release-ga-and-support-schedules and the go client should be in one of the versions listed here https://aerospike.com/developer/client/client_matrix to be  to be "fully compatible" with the server version.


With TLS and the current dependency, I successfully tested TLS with `nginx:latest` (as I shown in the comment of the PR) and with the following configuration with `AEROSPIKE_VERSION=ee-6.4.0.7_2` with and without mutual tls.


## Test Configuration

<details>
   <summary>Aerospike EE TLS Configuration</summary>

The aerospike configuration file that we will use (Note the "tls"settings). Read more in https://aerospike.com/docs/server/operations/configure/network/tls 

```properties
# This stanza must come first.
service {
  feature-key-file /etc/aerospike/features.conf
  cluster-name aerocluster
}

logging {
        console {
        context any info
    }
}

network {
  tls aerocluster {
    ca-file /opt/aerospike/etc/certs/ca.crt # used for client authentication
    cert-file /opt/aerospike/etc/certs/aerocluster.crt
    key-file /opt/aerospike/etc/certs/aerocluster.key
  }
  service {
    address any
    port 3000
    tls-address any
    tls-port 3003
    tls-name aerocluster
    tls-authenticate-client any # Set to false to disable Mutual TLS
  }

  heartbeat {
    mode mesh
    address local
    port 3002
    interval 150
    timeout 10
  }

  fabric {
    address local
    port 3001
  }

}

namespace test {
  default-ttl 30d # use 0 to never expire/evict.
  memory-size 1G
  nsup-period 120
  replication-factor 1
  storage-engine device {
    data-in-memory false # if true, in-memory, persisted to the filesystem
    file /opt/aerospike/data/test.dat
    filesize 4G
    read-page-cache true
  }
}
```

Docker-Compose File:

Tested with environment variable `AEROSPIKE_VERSION: ee-6.4.0.7_2, ee-5.7.0.17`

```yml
version: '2.3'

services:
  aerospike:
    image: docker.elastic.co/integrations-ci/beats-aerospike:${AEROSPIKE_VERSION:-3.9.0}-1
    build:
      context: ./_meta
      args:
        AEROSPIKE_VERSION: ${AEROSPIKE_VERSION:-3.9.0}
    volumes:
      # Needed for ee version below 6.x
      # - ./features.conf:/etc/aerospike/features.conf:ro
      - ./certs/:/opt/aerospike/etc/certs/:ro
      - ./aerospike-tls.conf:/opt/aerospike/etc/aerospike.conf:ro
      
    command:
      - "--config-file"
      - "/opt/aerospike/etc/aerospike.conf"
    ports:
      # TCP Plain Port
      - 3000:3000
      # TLS Port
      - 3003:3003
```

Metricbeat modules.d/aerospike.yml

```yml
- module: aerospike
  #metricsets:
  #  - namespace
  period: 10s
  hosts: ["localhost:3003"]

  # Aerospike Cluster Name
  cluster_name: aerocluster

  # Read more about SSL Configurations https://www.elastic.co/guide/en/beats/metricbeat/current/configuration-ssl.html
  ssl.enabled: true
  ssl.certificate_authorities: ["./certs/ca.crt"]
  ssl.certificate: "./certs/aeroclusterclient.crt"
  ssl.key: "./certs/aeroclusterclient.key"
```

</details>

File: metricbeat/module/aerospike/_meta/config.yml
Comment (leehinman) [inline metricbeat/module/aerospike/_meta/config.yml:11]: Sorry, I forgot about this last time.  But you need to add the new options to the `config.reference.yml` as well so it shows up in the `metricbeat.reference.yml` file.  And should add a link to the SSL options in the `docs.asciidoc` so the documentation gets updated as well.

**Review by leehinman — CHANGES_REQUESTED** [review]

File: metricbeat/module/aerospike/_meta/config.yml
Comment (herrBez) [reply metricbeat/module/aerospike/_meta/config.yml]: It makes sense :). I think I got it. I added to fields.yml the "ssl" settings and it has been generated automatically ^^. I took inspiration from this PR https://github.com/elastic/beats/commit/4db5f0aaefeb9038d7656b470a44a8cc37d19a37. I hope it looks better now!

**Comment by herrBez** [comment]
Hi there,
Are there any news?

**Comment by pierrehilbert** [comment]
Looks like you didn't ask for another review, I just triggered it

**Review by leehinman — APPROVED** [review]
LGTM

**Comment by herrBez** [comment]
Hi thank you very much for the feedback! I assume that in the meantime the ownership of the module changed to the team  elastic/obs-infraobs-integrations as a consequence of https://github.com/elastic/beats/pull/38249. Should we change the team-label?

**Comment by herrBez** [comment]
Hi, 
It is correct to select a team as an reviewer or should we address single developers?

**Comment by lalit-satapathy** [comment]
> @lalit-satapathy could we have someone for a review here?

CC: @gpop63 @shmsr

File: metricbeat/module/aerospike/aerospike_test.go
Comment (shmsr) [inline metricbeat/module/aerospike/aerospike_test.go:108]: Replace this with this:

https://github.com/elastic/beats/blob/8105935e24fb12a2120488290015a38f143d8471/x-pack/filebeat/input/salesforce/helper.go#L31

File: metricbeat/module/aerospike/_meta/config.reference.yml
Comment (shmsr) [inline metricbeat/module/aerospike/_meta/config.reference.yml:5]:
```suggestion
  hosts: ["localhost:3000"]
  
```

File: metricbeat/docs/modules/aerospike.asciidoc
Comment (shmsr) [inline metricbeat/docs/modules/aerospike.asciidoc:45]: Why cluster_name is now required? Also, irrespective of the return value of  `config.TLS.IsEnabled()`, the cluster name is set regardless. It was working previously without the cluster_name right?

**Review by shmsr — CHANGES_REQUESTED** [review]

File: metricbeat/module/aerospike/aerospike.go
Comment (gpop63) [inline metricbeat/module/aerospike/aerospike.go:31]: Do we need to use a pointer here for `ClusterName`? `string` would be cleaner. Also, is `ClusterName` somehow required when using TLS?

File: metricbeat/module/aerospike/namespace/namespace.go
Comment (gpop63) [inline metricbeat/module/aerospike/namespace/namespace.go:113]: I wonder if it's worth creating a small integration test for this method to check if SSL connection is successful. WDYT?  @shmsr

Comment (shmsr) [reply metricbeat/module/aerospike/namespace/namespace.go:113]: Yes. If it is simple to add, +1 to this.

File: metricbeat/module/aerospike/aerospike_test.go
Comment (shmsr) [inline metricbeat/module/aerospike/aerospike_test.go:158]:
```suggestion
					"Aerospike policy the error produced is not the one expected: got '%s', expected '%s'", err.Error(), test.expectedErr.Error())
```

File: metricbeat/module/aerospike/aerospike_test.go
Comment (shmsr) [inline metricbeat/module/aerospike/aerospike_test.go:111]:
```suggestion
```

**Review by shmsr — CHANGES_REQUESTED** [review]

File: metricbeat/docs/modules/aerospike.asciidoc
Comment (herrBez) [reply metricbeat/docs/modules/aerospike.asciidoc:45]: Hi, yes it did work before. By further reading the documentation of the Aerospike the ClusterName setting is actually used for:

> Expected cluster name. If populated, the clusterName must match the cluster-name field in the service section in each server configuration. This ensures that the specified seed nodes belong to the expected cluster on startup. If not, the client will refuse to add the node to the client's view of the cluster.

So it is not necessarily related to TLS. Indeed, the example of the aerospike documentation with TLS and Java Client https://aerospike.com/docs/server/operations/configure/network/tls/mtls_java does not use the parameter.

So I think, it is okay if the setting is set regardless of TLS. I am open to remove it from the PR if we want to keep the scope of the PR tight.

File: metricbeat/module/aerospike/aerospike.go
Comment (herrBez) [reply metricbeat/module/aerospike/aerospike.go:31]: As commented here https://github.com/elastic/beats/pull/38126#discussion_r1555338419 this setting is not actually required to make TLS work. It's only needed to avoid that a client connects to a different/unexpected cluster. 

I think we can get rid of the setting, if we want to make the PR only focus on TLS

File: metricbeat/module/aerospike/namespace/namespace.go
Comment (herrBez) [reply metricbeat/module/aerospike/namespace/namespace.go:113]: Hi,

In this comment https://github.com/elastic/beats/pull/38126#issuecomment-1999489894 I added the docker-compose and TLS configuration I used to test TLS. Let me know if you have any question.

The only problem I see is that the original test/expected version of the server is 3.9.0, which is very old and I am not sure it supports TLS/it is worth the hassle of finding out how to configure it (we may use the NGINX version I provided in the original comment to avoid relying on aerospike-version).

BTW Using a modern version like `AEROSPIKE_VERSION=ee-6.4.0.7_2` (ee = enterprise-edition) should do the trick.

Comment (shmsr) [reply metricbeat/module/aerospike/namespace/namespace.go:113]: Yeah, let's update it. Anyway, the current version 3.9 is very old (you even mentioned).

File: metricbeat/docs/modules/aerospike.asciidoc
Comment (shmsr) [reply metricbeat/docs/modules/aerospike.asciidoc:45]: No, it is fine. Let's keep it.

File: x-pack/metricbeat/metricbeat.reference.yml
Comment (shmsr) [inline x-pack/metricbeat/metricbeat.reference.yml:164]:
```suggestion
  # Optional SSL/TLS (disabled by default)
```

Comment (shmsr) [reply x-pack/metricbeat/metricbeat.reference.yml:164]: Make similar changes everywhere in this PR.

**Review by shmsr — CHANGES_REQUESTED** [review]

File: x-pack/metricbeat/metricbeat.reference.yml
Comment (herrBez) [reply x-pack/metricbeat/metricbeat.reference.yml:164]: Resolved with https://github.com/elastic/beats/pull/38126/commits/4a15b5bcb10643bc3e5acafaae0d46a3b9eda521

**Comment by shmsr** [comment]
Are we adding the tests that @gpop63 suggested? If yes, then I'll wait for the changes and then approve. If no, then I'll approve this PR right away; changes look good.

cc: @gpop63

**Review by gpop63 — APPROVED** [review]
LGTM, make sure to fix the docs and x-pack/metricbeat CI. `make update` should probably fix them.

**Comment by herrBez** [comment]
Thank you all for the help and the patience! After the merge I noticed we did not add an entry in the release notes. Should we add one?

**Comment by shmsr** [comment]
> Thank you all for the help and the patience! After the merge I noticed we did not add an entry in the release notes. Should we add one?

Yes, we should! Thanks.

**Comment by herrBez** [comment]
Hi, how should we proceed? By opening a new PR? I tried committing in my branch but this PR did not get the update.

(BTW the release notes seems to contain duplicated sections e.g line [243](https://github.com/elastic/beats/blob/main/CHANGELOG.next.asciidoc?plain=1#L243) and [245](https://github.com/elastic/beats/blob/main/CHANGELOG.next.asciidoc?plain=1#L245) . Is it expcted?)

**Comment by shmsr** [comment]
> Hi, how should we proceed? By opening a new PR? I tried committing in my branch but this PR did not get the update.
> 
> (BTW the release notes seems to contain duplicated sections e.g line [243](https://github.com/elastic/beats/blob/main/CHANGELOG.next.asciidoc?plain=1#L243) and [245](https://github.com/elastic/beats/blob/main/CHANGELOG.next.asciidoc?plain=1#L245) . Is it expcted?)

No. Duplicates are not expected. Also, yes you can create a new PR to include the changelog entry. Please do it ASAP and tomorrow we have code freeze for 8.14 release.

**Comment by herrBez** [comment]
Done in https://github.com/elastic/beats/pull/38923 .

**Pull request merged**
